core/nixpkgs.bzl

```
@@ -903,3 +884,92 @@ def nixpkgs_flake_package(
     )
 
     _nixpkgs_flake_package(**kwargs)
+
+def _run_nix_shell_impl(ctx, **kwargs):
+    out_file = ctx.actions.declare_file("%s" % ctx.attr.output)
+
+    nix_env = ctx.file.stdenv
+
+    ctx.actions.run_shell(
+      inputs = [nix_env],
+      outputs = [out_file],
+      progress_message = "Running nix shell...",
+      arguments = [out_file.path],
+      command = """
+        source {}
+        {}
+        """.format(nix_env.path, ctx.attr.cmd, out_file.path)
+    )
+    return [DefaultInfo(files = depset([out_file]))]
+
+run_nix_shell = rule(
+    implementation = _run_nix_shell_impl,
+    attrs = {
+        "output": attr.string(
```

```suggestion
        "outputs": attr.output_list(
```

I think an [output list](https://bazel.build/rules/lib/toplevel/attr#output_list) would give us a bit more flexibility. This way, users can define multiple outputs for a single action. This also avoids the need to `ctx.actions.declare_file`. Instead, the file is available through `ctx.outputs.outputs`.